docs: update developer reference docs (#126)
matter-labs/zksync-docs: content/00.build/65.developer-reference/50.era-contracts/60.elliptic-curve-precompiles.md

These are a special kind of contracts that are bundled with the EVM at fixed addresses and can be called with a determined gas cost.
The addresses start from 1, and increment for each contract.
New hard forks may introduce new precompiled contracts.
They are called from the opcodes like regular contracts, with instructions like CALL.
The gas cost mentioned here is purely the cost of the contract
and does not consider the cost of the call itself nor the instructions to put the parameters in memory.

For Go-Ethereum, the code being run is written in Go, and the gas costs are defined in each precompile spec.

In the case of ZKsync Era, ecAdd and ecMul precompiles are written as a smart contract for two reasons:

- zkEVM needs to be able to prove their execution (and at the moment it cannot do that if the code being run is executed outside the VM).
- Writing custom circuits for Elliptic curve operations is hard, and time-consuming, and after all such code is harder to maintain and audit.

## Field Arithmetic

The BN254 (also known as alt-BN128) is an elliptic curve defined by the equation $y^2 = x^3 + 3$ over the finite field $\mathbb{F}_p$,

> being $p = 21888242871839275222246405745257275088696311157297823662689037894645226208583.

Reviewer: Good catch. Yes, I agree. It should have a closing `$`

Then the compiler performs long division for reduction (but only the remainder is kept),
in the generic form it is an expensive operation and costs many opcode executions,
which can’t be compared to the cost of one opcode execution.
The worst thing is that `mulmod` is used a lot for the modulo inversion, so optimizing this one opcode gives a huge benefit to the precompiles.

### Multiplication

As said before, multiplication was carried out by implementing the Montgomery reduction,
which works with general moduli and provides a significant speedup compared to the naïve approach.

The squaring operation is obtained by multiplying a number by itself.
However, this operation can have an additional speedup by implementing the SOS Montgomery squaring.

### Inversion

Inversion was performed using the extended binary Euclidean algorithm (also known as extended binary greatest common divisor).
This algorithm is a modification of Algorithm 3 `MontInvbEEA` from [Montgomery inversion](https://cetinkayakoc.net/docs/j82.pdf).

### Exponentiation

The exponentiation was carried out using the square and multiply algorithm, which is a standard technique for this operation.

## Montgomery Form

> Let’s take a number `R`, such that `gcd(N, R) == 1` and `R` is a number by which we can efficiently divide and take module over it

Reviewer 2: This occurs in a few places, but wasn't sure if we meant `modulus` (e.g. the modulus `N`) vs `modulo` or `Module` (the Algebraic structure between a Lattice/Ring and an Algebra)

> Let’s take a number `R`, such that `gcd(N, R) == 1` and `R` is a number by which we can efficiently divide and take module over it

Reviewer: I think, it should be `modulo` if we are talking about operations and `modulus` to refer to the number itself.

(for example power of two or better machine word, aka 2^256).
Then transform every number to the form of `x * R mod N` / `y * R mod N` and then we get efficient modulo addition and multiplication.
The only thing is that before working with numbers we need to transform them to the form from `x mod N`  to the `x * R mod N`
and after performing operations transform the form back.

For the latter, we will assume that `N` is the module that we use in computations,

> and `R` is $2^{256}$, since we can efficiently divide and take module over this number

Reviewer: ```suggestion
For the latter, we will assume that `N` is the module that we use in computations,
and `R` is $2^{256}$, since we can efficiently divide and take modulo over this number
```

> For the latter, we will assume that `N` is the module that we use in computations,
> and `R` is $2^{256}$, since we can efficiently divide and take module over this number
> and it practically satisfies the property of `gcd(N, R) == 1`.

Reviewer: We can correct the instances of `module` here:
```suggestion
For the latter, we will assume that `N` is the modulus that we use in computations,
and `R` is $2^{256}$, since we can efficiently divide and take modulo over this number
and it practically satisfies the property of `gcd(N, R) == 1`.
```

/// @return invmod The result of the Montgomery modular inverse (in Montgomery form).
function montgomeryModularInverse(a) -> invmod {
    invmod := binaryExtendedEuclideanAlgorithm(a)
}
```

As said before, we use a modified version of the bEE algorithm that lets us “skip” the Montgomery reduction step.

The regular algorithm would be $REDC((aR \mod N)^{−1}(R^3 \mod N))$ which involves a regular inversion plus a multiplication by a value that can be precomputed.

## ECADD

Precompile for computing elliptic curve point addition. The points are represented in affine form, given by a pair of coordinates $(x,y)$.

Affine coordinates are the conventional way of expressing elliptic curve points, which use 2 coordinates. The math is concise and easy to follow.

For a pair of constants $a$ and $b$, an elliptic curve is defined by the set of all points $(x,y)$
that satisfy the equation $y^2=x^3+ax+b$, plus a special “point at infinity” named $O$.

### Point Doubling

To compute $2P$ (or $P+P$), there are three cases:

- If $P = O$, then $2P = O$.
- Else $P = (x, y)$
  - If $y = 0$, then $2P = O$.
  - Else $y≠0$, then

>         $$

Reviewer 2: I don't think having tabs before LaTeX (`$$`) is supported:
![image](https://github.com/matter-labs/zksync-docs/assets/1890113/0f891675-5ab7-4167-b510-7e5e0af94ccd)

Reviewer: Yeah, we should remove it then

        \begin{gather*} \lambda = \frac{3x_{p}^{2} + a}{2y_{p}} \\ x_{r} = \lambda^{2} - 2x_{p} \\ y_{r} = \lambda(x_{p} - x_{r}) - y_{p}\end{gather*}
        $$

The complicated case involves approximately 6 multiplications, 4 additions/subtractions, and 1 division.
There could also be 4 multiplications, 6 additions/subtractions, and 1 division,
and if you want you could trade a multiplication with 2 more additions.

### Point Addition

To compute $P + Q$ where $P \neq Q$, there are four cases:

- If $P = 0$ and $Q \neq 0$, then $P + Q = Q$.
- If $Q = 0$ and $P \neq 0$, then $P + Q = P$.
- Else $P = (x_{p},\ y_{p})$ and$Q = (x_{q},\ y_{q})$
  - If $x_{p} = x_{q}$ (and necessarily $y_{p} \neq y_{q}$), then $P + Q = O$.
  - Else $x_{p} \neq x_{q}$, then

        $$
        \begin{gather*} \lambda = \frac{y_{2} - y_{1}}{x_{2} - x_{1}} \\ x_{r} = \lambda^{2} - x_{p} - x_{q} \\ y_{r} = \lambda(x_{p} - x_{r}) - y_{p}\end{gather*}
        $$

>         and $P + Q = R = (x_{r},\ y_{r})$.

Reviewer 2: Another indention that I don't believe we support:
![image](https://github.com/matter-labs/zksync-docs/assets/1890113/1306a277-27d8-4c62-a017-c69f374fcd9d)

Reviewer: Yeah, we can remove the indentations.

For the interesting case where $P = (X_{p},\ Y_{p},\ Z_{p})$ , $Q = (X_{q},\ Y_{q},\ Z_{q})$, and $X_{p}Z_{q} ≠ X_{q}Z_{p}$,
let’s convert the affine arithmetic to projective arithmetic.

After expanding and simplifying the equations
([demonstration here](https://www.nayuki.io/page/elliptic-curve-point-addition-in-projective-coordinates)),
the following substitutions come out

$$
\begin{align*}
T_{0} &= Y_{p}Z_{q}\\
T_{1} &= Y_{q}Z_{p}\\
T &= T_{0} - T_{1}\\
U_{0} &= X_{p}Z_{q}\\
U_{1} &= X_{q}Z_{p}\\
U &= U_{0} - U_{1}\\
U_{2} &= U^{2}\\
V &= Z_{p}Z_{q}\\
W &= T^{2}V−U_{2}(U_{0}+U_{1}) \\
\end{align*}
$$

Using them, we can write

$$
\begin{align*} X_{r}  &= UW \\ Y_{r} &= T(U_{0}U_{2}−W)−T_{0}U^{3} \\ Z_{r} &= U^{3}V \end{align*}
$$

As we can see, the complicated case involves approximately 15 multiplications, 6 additions/subtractions, and 0 divisions.

## P256

The P256Verify smart contract emulates the [RIP-7212](https://github.com/ethereum/RIPs/blob/master/RIPS/rip-7212.md)'s `P256VERIFY` precompile,
enabling ECDSA signature verification using the secp256r1 curve (P-256) within the zkEVM environment.
This contract acts as a wrapper for the associated circuit.

> It includes helper functions like unsafePackPrecompileParams to pack precompile parameters

Reviewer 2: suggestion: Add code backtick for readability of function names
```suggestion
It includes helper functions like `unsafePackPrecompileParams` to pack precompile parameters
```

(same for `precompileCall` one line below)
